WordPressCom-Stats-iOS/UI/WPStatsViewController.m

```diff
@@ -49,18 +60,36 @@ - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
         self.statsTableViewController = tableVC;
         tableVC.statsDelegate = self.statsDelegate;
         tableVC.statsProgressViewDelegate = self;
-        tableVC.statsService = [[WPStatsService alloc] initWithSiteId:self.siteID siteTimeZone:self.siteTimeZone oauth2Token:self.oauth2Token andCacheExpirationInterval:5 * 60];
-;
+        tableVC.statsService = [self serviceForSiteID:self.siteID];
     } else if ([segue.identifier isEqualToString:@"InsightsTableEmbed"]) {
         InsightsTableViewController *insightsTableViewController = (InsightsTableViewController *)segue.destinationViewController;
         self.insightsTableViewController = insightsTableViewController;
-        insightsTableViewController.statsService = [[WPStatsService alloc] initWithSiteId:self.siteID siteTimeZone:self.siteTimeZone oauth2Token:self.oauth2Token andCacheExpirationInterval:5 * 60];
+        insightsTableViewController.statsService = [self serviceForSiteID:self.siteID];
         insightsTableViewController.statsProgressViewDelegate = self;
         insightsTableViewController.statsTypeSelectionDelegate = self;
         insightsTableViewController.statsDelegate = self.statsDelegate;
     }
 }
 
+#pragma mark - Convenience methods
+- (nonnull WPStatsService *)serviceForSiteID:(nonnull NSNumber *)siteID
+{
+    NSParameterAssert(siteID);
+    
+    WPStatsService *service;
+    if (self.statsServiceCache) {
+        service = [self.statsServiceCache serviceForSiteID:siteID];
+    }
+    if (!service) {
+        if ((service = [[WPStatsService alloc] initWithSiteId:self.siteID siteTimeZone:self.siteTimeZone oauth2Token:self.oauth2Token andCacheExpirationInterval:5 * 60])) {
```

Comment:
I think that inlining the equality check and the assignment here makes this code harder to read and runs the risk of introducing subtle errors. I'd split this apart.